[IMP] orm, web: add HTML properties

std-odoo opened the pull request:
Purpose
=======
Add HTML properties, to be able to format text in the properties field.

Technical
=========
The default value is sanitized in the properties definition in `convert_to_column` /
`convert_to_cache`, like it's done for standard HTML field. However, the properties
value is sanitized in `convert_to_read`. The reason is that we need the property type
to know that we need to sanitize it, and writing a dict is a supported feature of the
properties. Because `convert_to_record` will call the read methods, it should not be
possible to get a non-sanitized value.

This is also very useful, because we can technically change the property type,
without changing the name (so if it was done in the same way as HTML filed, one could
have created a text property, and then change the type on the parent).

Task-4712395

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/208718.png)](https://mergebot.odoo.com/odoo/odoo/pull/208718)

std-odoo commented:
@ryv-odoo Hi :) 

FYI, we will add this to properties :) (nothing urgent to review, but it would be nice to get a small review when you have the time, the target is 18.4 )

HydrionBurst commented on odoo/orm/fields_properties.py:
```diff
@@ -518,6 +518,9 @@ def _parse_json_types(cls, values_list, env, res_ids_per_model):
                 ] if res_model in env else []
 
+            elif property_type == 'html':
+                property_value = html_normalize(html_sanitize(property_value))
```
Is the `html_normalize` necessary here?

std-odoo replied:
I took the same logic as standard HTML field for consistency

HydrionBurst commented on odoo/orm/fields_properties.py:
```diff
@@ -946,6 +949,9 @@ def _validate_properties_definition(self, properties_definition, env):
                 raise ValueError(f'Wrong property type {property_type!r}.')
 
+            if property_type == 'html' and (default := property_definition.get('default')):
+                property_definition['default'] = html_sanitize(default)
```
Not sure but it would be good to recheck the parameters of the `html_sanitize`
`field.Html.sanitize_attributes is True` by default, but `html_sanitize`'s default parameter is `sanitize_attributes=False`

HydrionBurst commented on odoo/addons/test_new_api/tests/test_properties.py:
```diff
@@ -788,6 +788,71 @@ def test_properties_field_read(self):
+                "definition_changed": True,
+            },
+        ]
+        self.assertEqual(self.message_2.attributes['text'], xss_payload)
```
I guess it is not a good practice to allow/encourage developers to call properties fields like this
Because after https://github.com/odoo/odoo/pull/205139. `self.message_2.attributes` returns a `Property` object whose `__getitem__` will call `convert_to_read` not in batch
`convert_to_read` will call `_get_res_ids_per_model` to check the `exists` of relevant records
if property has relational attributes, and `self.message_2.attributes['text']` is used. It would be one query per field read. The performance might be a problem in the future.

ryv-odoo reviewed (commented):
I don't have a lot to say in this particular PR.

IMHO, we are still in the wrong direction with Properties Field. It looks like we are currently re-implementing the ORM and every of standard Fields for no good reason :cry: 

Note that the "placeholder" buttons of the html editor doesn't work for html property.

std-odoo commented:
@ryv-odoo Hi :) 

Thanks for the review, as discussed on discord I removed the sanitization for `web_read_group` and `read_group`, and I disabled the group by in the client side

Placeholder button are also fixed, about the re-implementing for each type, it's already the case for other properties type no ?

ryv-odoo commented:
Thanks for the change @std-odoo .

> Placeholder button are also fixed, about the re-implementing for each type, it's already the case for other properties type no ?

Indeed, my point stands for the all the Properties features not only for this new type.
